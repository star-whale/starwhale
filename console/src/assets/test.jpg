version https://git-lfs.github.com/spec/v1
oid sha256:b030429eeb8453458a00d82cfe52ec4611c4d6c2bb306a5e61cc0946de9948dd
size 97175
