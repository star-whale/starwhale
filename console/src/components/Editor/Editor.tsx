import React, { useMemo, useRef } from 'react'
import EditorContextProvider, { StoreType } from '@starwhale/core/context/EditorContextProvider'
import { createCustomStore } from '@starwhale/core/store'
import WidgetRenderTree from '@starwhale/core/widget/WidgetRenderTree'
import { EventBusSrv } from '@starwhale/core/events'
import { useJob } from '@/domain/job/hooks/useJob'
import { tablesOfEvaluation, WidgetStateT, WidgetStoreState } from '@starwhale/core'
import BusyPlaceholder from '@starwhale/ui/BusyLoaderWrapper/BusyPlaceholder'
import { tranformState } from './utils'
import { useProject } from '@project/hooks/useProject'

export function withProject(EditorApp: React.FC) {
    return function EditorLoader(props: any) {
        const { project } = useProject()
        const projectId = project?.id
        const { job } = useJob()
        // eslint-disable-next-line prefer-template
        const prefix = projectId && job?.uuid ? tablesOfEvaluation(projectId, job?.uuid) + '/' : undefined
        const storeKey = job?.modelName ? ['evaluation-model', job?.modelName].join('-') : undefined
        if (!prefix || !storeKey || !projectId) {
            return <BusyPlaceholder type='spinner' />
        }

        const dynamicVars = { prefix, storeKey, projectId }

        return <EditorApp {...props} dynamicVars={dynamicVars} />
    }
}

const empty = {}

export function witEditorContext<EditorAppPropsT>(EditorApp: React.FC<EditorAppPropsT>, rawState: WidgetStateT) {
    return function EditorContexted(props: EditorAppPropsT & { dynamicVars?: any }) {
        const state = useMemo(() => tranformState(rawState), [])
        const store = useRef<StoreType>()
        const value = useMemo(() => {
            console.log('raw')
            if (rawState) {
                store.current = createCustomStore(state)
            }
            const eventBus = new EventBusSrv()
            return {
                store: store.current,
                eventBus,
            }
        }, [state])

        return (
            <EditorContextProvider
                value={{
                    ...value,
                    dynamicVars: props.dynamicVars || empty,
                }}
            >
                <EditorApp {...props} />
            </EditorContextProvider>
        )
    }
}

const initialState = {
    key: 'widgets',
    tree: [
        {
            type: 'ui:dndList',
            children: [
                {
                    type: 'ui:section',
                },
            ],
        },
    ],
    widgets: {},
    defaults: {},
}
const Editor = withProject(witEditorContext(WidgetRenderTree, initialState))

export { Editor, initialState }
export default Editor
