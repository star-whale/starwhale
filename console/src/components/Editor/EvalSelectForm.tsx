import React, { useImperativeHandle } from 'react'
import { GridTable } from '@starwhale/ui/GridTable'
import ProjectSelector from '@/domain/project/components/ProjectSelector'
import { ColumnSchemaDesc, tableNameOfSummary, useFetchDatastoreByTable } from '@starwhale/core/datastore'
import { ITableState, createCustomStore } from '@starwhale/ui/GridTable/store'
import shallow from 'zustand/shallow'
import useDatastorePage from '@starwhale/core/datastore/hooks/useDatastorePage'
import { useDatastoreSummaryColumns } from '@starwhale/ui/GridDatastoreTable/hooks/useDatastoreSummaryColumns'
import GridCombineTable from '@starwhale/ui/GridTable/GridCombineTable'
import { useEvent } from 'react-use'
import { useEventCallback } from '@starwhale/core'

const selector = (s: ITableState) => ({
    rowSelectedIds: s.rowSelectedIds,
    currentView: s.currentView,
    initStore: s.initStore,
    getRawConfigs: s.getRawConfigs,
    onCurrentViewIdChange: s.onCurrentViewIdChange,
    getRawIfChangedConfigs: s.getRawIfChangedConfigs,
})

export const useStore = createCustomStore('add-evaluation', {}, false)

function EvalProjectList({
    projectId,
    onRowSelectedChange,
}: {
    projectId: string
    onRowSelectedChange: (
        data: Record<
            string,
            {
                records: Record<string, any>[]
                columnTypes: ColumnSchemaDesc[]
            }
        >
    ) => void
}) {
    const summaryTableName = tableNameOfSummary(projectId)

    const { currentView } = useStore(selector, shallow)

    const { page, setPage, params } = useDatastorePage({
        pageNum: 1,
        pageSize: 999,
        sortBy: currentView?.sortBy || 'sys/id',
        sortDirection: currentView?.sortBy ? (currentView?.sortDirection as any) : 'DESC',
        queries: currentView?.queries,
        tableName: summaryTableName,
    })
    const { columnTypes, records } = useFetchDatastoreByTable(params, !!projectId)

    const $columns = useDatastoreSummaryColumns(columnTypes as any, {
        projectId,
    })

    const handelRowSelectedChange = useEventCallback((ids: any[]) => {
        const rows = ids.map((id) => records.find((r) => r.id?.value === id)) as any

        onRowSelectedChange({
            [summaryTableName]: {
                records: rows,
                columnTypes,
            },
        })
    })

    return (
        <GridCombineTable
            paginationable
            queryable
            columnable
            compareable={false}
            page={page}
            onPageChange={setPage}
            store={useStore}
            records={records}
            columnTypes={columnTypes}
            columns={$columns}
            onRowSelectedChange={handelRowSelectedChange}
        />
    )
}

const EvalSelectForm = React.forwardRef((props, ref: any) => {
    const [projectId, setProjectId] = React.useState<string>('')
    const [projectItem, setProjectItem] = React.useState<any>(null)
    const [rows, setRows] = React.useState<
        Record<
            string,
            {
                records: Record<string, any>[]
                columnTypes: ColumnSchemaDesc[]
            }
        >
    >({})

    useImperativeHandle(
        ref,
        () => ({
            getData() {
                return { projectId, projectItem, tableMap: rows }
            },
        }),
        [projectId, projectItem, rows]
    )

    console.log(rows)

    return (
        <div className='flex flex-column gap-12px'>
            <div className='w-280px'>
                <ProjectSelector
                    value={projectId}
                    onChange={(id) => setProjectId(id)}
                    onChangeItem={(item) => setProjectItem(item)}
                />
            </div>
            <div className='h-380px w-full'>
                <EvalProjectList
                    projectId={projectId}
                    onRowSelectedChange={(r: any) =>
                        setRows((prev: any) => ({
                            ...prev,
                            ...r,
                        }))
                    }
                />
            </div>
        </div>
    )
})

export { EvalSelectForm }
export default EvalSelectForm
