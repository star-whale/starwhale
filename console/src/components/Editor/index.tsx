import React, { useEffect, useMemo } from 'react'
import EditorContextProvider from '@starwhale/core/context/EditorContextProvider'
import { registerWidgets } from '@starwhale/core/widget/WidgetFactoryRegister'
import { createCustomStore } from '@starwhale/core/store'
import WidgetRenderTree from '@starwhale/core/widget/WidgetRenderTree'
import { EventBusSrv } from '@starwhale/core/events'
import { useProject } from '@/domain/project/hooks/useProject'
import { useJob } from '@/domain/job/hooks/useJob'
import { tablesOfEvaluation } from '@starwhale/core'
import { useParams } from 'react-router'
import BusyPlaceholder from '../BusyLoaderWrapper/BusyPlaceholder'
import { tranformState } from './utils'

registerWidgets()

export function withEditorRegister(EditorApp: React.FC) {
    return function EditorLoader(props: any) {
        // const [registred, setRegistred] = React.useState(false)
        // useEffect(() => {
        //     // registerRemoteWidgets().then((module) => {
        //     //     setRegistred(true)
        //     // })
        // }, [])
        // if (!registred) {
        //     return <BusyPlaceholder type='spinner' />
        const [module, setModule] = React.useState<any>(null)
        // }
        useEffect(() => {
            // @ts-ignore
            // eslint-disable-next-line
            import('http://127.0.0.1:8080/widget.js').then((module) => {
                // setRegistred(true)
                console.log(module)
                setModule(module)
            })
        }, [])

        // log.debug('WidgetFactory', WidgetFactory.widgetMap)
        // @FIXME
        const { projectId } = useParams<{ projectId: string }>()
        const { project } = useProject()
        const { job } = useJob()
        // eslint-disable-next-line prefer-template
        const prefix = project?.name && job?.uuid ? tablesOfEvaluation(project?.name, job?.uuid) + '/' : undefined
        const storeKey = job?.modelName ? ['evaluation-model', job?.modelName].join('-') : undefined
        if (!prefix || !storeKey || !projectId) {
            return <BusyPlaceholder type='spinner' />
        }

        const dynamicVars = { prefix, storeKey, projectId }

        return <EditorApp {...props} dynamicVars={dynamicVars} />
    }
}

export function witEditorContext(EditorApp: React.FC, rawState: typeof initialState) {
    return function EditorContexted(props: any) {
        const state = useMemo(() => tranformState(rawState), [])
        // @NOTICE must only init once
        const value = useMemo(() => {
            // @ts-ignore
            const store = createCustomStore(state)
            const eventBus = new EventBusSrv()
            return {
                store,
                eventBus,
            }
        }, [state])

        return (
            <EditorContextProvider
                value={{
                    ...value,
                    dynamicVars: props.dynamicVars,
                }}
            >
                <EditorApp {...props} />
            </EditorContextProvider>
        )
    }
}

const initialState = {
    key: 'widgets',
    tree: [
        {
            type: 'ui:dndList',
            children: [
                {
                    type: 'ui:section',
                },
            ],
        },
    ],
    widgets: {},
    defaults: {},
}
const Editor = withEditorRegister(witEditorContext(WidgetRenderTree, initialState))

export default Editor
