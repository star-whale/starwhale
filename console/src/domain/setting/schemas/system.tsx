import { IPrivileges } from '@/api/const'
import { IResourceSchema } from '@/domain/base/schemas/resource'

export type IDeviceSchema = IResourceSchema

export interface ISystemVersionSchema {
    id: string
    version: string
}

export interface IAgentSchema {
    connectedTime: number
    id: string
    ip: string
    serialNumber: string
    status: string
    version: string
}

export type ISystemSettingSchema = string

export type ISystemResource = {
    name: string
    max: number
    min: number
    defaults: number
}
export type ISystemResourcePool = {
    name: string
    nodeSelector: Record<string, string>
    resources: ISystemResource[]
}

export type ISystemFeaturesSchema = {
    disabled?: Array<keyof IPrivileges>
}
