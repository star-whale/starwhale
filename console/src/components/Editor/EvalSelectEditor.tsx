import React from 'react'
import { witEditorContext } from './Editor'
import { WidgetRenderTree, WidgetRenderTreePropsT } from '@starwhale/core/widget'
import { WidgetStateT } from '@starwhale/core'

const initialState: WidgetStateT = {
    key: 'widgets',
    tree: [
        {
            type: 'ui:section',
            optionConfig: {
                isEvaluationList: true,
            },
        },
    ],
    widgets: {},
    defaults: {},
}

const EvalSelectEditor = witEditorContext<WidgetRenderTreePropsT>(WidgetRenderTree, initialState)

export { EvalSelectEditor }
export default EvalSelectEditor
