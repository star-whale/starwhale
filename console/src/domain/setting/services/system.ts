import { IListQuerySchema, IListSchema } from '@/domain/base/schemas/list'
import axios from 'axios'
import { IAgentSchema, ISystemVersionSchema } from '../schemas/system'

export async function fetchSystemVersion(): Promise<ISystemVersionSchema> {
    const resp = await axios.get('/api/v1/system/version')
    return resp.data
}

export async function listAgents(query: IListQuerySchema): Promise<IListSchema<IAgentSchema>> {
    const resp = await axios.get<IListSchema<IAgentSchema>>('/api/v1/system/agent', {
        params: query,
    })
    return resp.data
}

export async function deleteAgent(serialNumber: string): Promise<any> {
    const resp = await axios.delete(`/api/v1/system/agent?serialNumber=${serialNumber}`)
    return resp.data
}
