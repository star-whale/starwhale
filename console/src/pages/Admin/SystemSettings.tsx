import React, { useCallback, useState } from 'react'
import { createForm } from '@/components/Form'
import useTranslation from '@/hooks/useTranslation'
import { Button, SIZE, KIND } from 'baseui/button'
import Editor from '@monaco-editor/react'
import yaml from 'js-yaml'
import { createUseStyles } from 'react-jss'
import { toaster } from 'baseui/toast'
import { useFetchSystemSetting } from '@/domain/setting/hooks/useSettings'
import { updateSystemSetting } from '@/domain/setting/services/system'

const { Form, FormItem, useForm } = createForm<any>()

const useStyles = createUseStyles({
    buttonGroup: {
        display: 'flex',
        gap: 20,
        marginTop: 60,
    },
})

export default function SystemSettings() {
    const styles = useStyles()
    const [t] = useTranslation()
    const [form] = useForm()
    const systemSetting = useFetchSystemSetting()
    const [loading, setLoading] = useState(false)
    const [values, setValues] = React.useState<any>(null)

    React.useEffect(() => {
        if (systemSetting.isSuccess) {
            if (typeof systemSetting.data === 'string')
                form.setFieldsValue({
                    setting: systemSetting.data,
                })
        }
    }, [systemSetting, form])

    const handleValuesChange = useCallback((_changes, values_) => {
        setValues(values_)
    }, [])

    const handleFinish = useCallback(
        async (values_: any) => {
            try {
                yaml.load(values_.setting)
            } catch (e) {
                toaster.negative(t('wrong yaml syntax'), { autoHideDuration: 1000 })
                throw e
            }
            setLoading(true)
            try {
                await updateSystemSetting(values_.setting)
            } finally {
                setLoading(false)
            }
        },
        [t]
    )

    return (
        <Form form={form} initialValues={values} onFinish={handleFinish} onValuesChange={handleValuesChange}>
            <FormItem label='' name='setting' required>
                <Editor height='500px' width='960px' defaultLanguage='yaml' theme='vs-dark' />
            </FormItem>
            <FormItem>
                <div className={styles.buttonGroup}>
                    <Button
                        size={SIZE.compact}
                        kind={KIND.secondary}
                        type='button'
                        onClick={() => {
                            form.setFieldsValue({
                                setting: systemSetting.data,
                            })
                        }}
                    >
                        {t('Reset')}
                    </Button>
                    <Button size={SIZE.compact} isLoading={loading}>
                        {t('Update')}
                    </Button>
                    <div style={{ flexGrow: 1 }} />
                </div>
            </FormItem>
        </Form>
    )
}
