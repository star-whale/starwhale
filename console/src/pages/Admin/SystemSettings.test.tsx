import * as React from 'react'
import {
    queryAllByText,
    fireEvent,
    getByText,
    screen,
    waitFor,
    queryAllByAttribute,
    render,
} from '@testing-library/react'
// import { queryByAttribute } from '@testing-library/dom'
import SystemSettings from './SystemSettings'
import { routeRender } from '@/test/provider'
import Editor from '@monaco-editor/react'

import * as system from '@/domain/setting/services/system'

jest.spyOn(system, 'fetchSystemSetting').mockImplementation(() =>
    Promise.resolve(`
        ---
        dockerSetting:
        registry: "abcd.com"
        `)
)

describe('SystemSettings', () => {
    describe('editor', () => {
        it('render', () => {
            const { container } = render(<Editor height='500px' width='960px' defaultLanguage='yaml' theme='vs-dark' />)
            expect(container).toMatchSnapshot()
        })
    })

    it('render', async () => {
        const { container } = routeRender(<SystemSettings />)
        expect(queryAllByText(container, 'Reset').length).toBe(1)
        expect(queryAllByText(container, 'Update').length).toBe(1)
    })

    it('render with fetchSystemSetting', async () => {
        const { container } = routeRender(<SystemSettings />)
        // expect(container).toMatchSnapshot()
        // await expect(document.querySelector('.monaco-editor')).toBeTruthy()
        // await new Promise(process.nextTick)
        await expect(document.querySelector('.formItem')).toBeTruthy()
        // await waitFor(() => expect(queryByAttribute('class', document.body, 'monaco-editor')).toBeTruthy())
        // await waitFor(() => expect(document.querySelector('.monaco-editor')).toBeTruthy(), {
        //     timeout: 1000,
        // })
        await waitFor(() => expect(system.fetchSystemSetting).toHaveBeenCalled())
    })
})
