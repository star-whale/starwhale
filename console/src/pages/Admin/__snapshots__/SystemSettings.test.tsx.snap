// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SystemSettings editor render 1`] = `
<div>
  <section
    style="display: flex; position: relative; text-align: initial; width: 960px; height: 500px;"
  >
    <div
      style="display: flex; height: 100%; width: 100%; justify-content: center; align-items: center;"
    >
      Loading...
    </div>
    <div
      style="width: 100%; display: none;"
    />
  </section>
</div>
`;

exports[`SystemSettings render with fetchSystemSetting 1`] = `
<div>
  <div
    class=""
  >
    <form>
      <div
        class="formItem"
      >
        <div
          class="ae af"
          data-baseweb="form-control-container"
        >
          <section
            style="display: flex; position: relative; text-align: initial; width: 960px; height: 500px;"
          >
            <div
              style="display: flex; height: 100%; width: 100%; justify-content: center; align-items: center;"
            >
              Loading...
            </div>
            <div
              style="width: 100%; display: none;"
            />
          </section>
        </div>
      </div>
      <div
        class="formItem"
      >
        <div
          class="ae af"
          data-baseweb="form-control-container"
        >
          <div
            class="buttonGroup-0-2-1"
            value="[object Object]"
          >
            <button
              class="ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az b0 b1 b2 b3 b4 b5 b6 b7 b8 b9 ba bb bc bd be bf bg bh bi bj bk bl bm"
              data-baseweb="button"
              type="button"
            >
              Reset
            </button>
            <button
              class="ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az b0 b1 b2 b3 b4 b5 b6 b7 b8 b9 ba bb bc bd be bf bg bh bi bn bo bp bq"
              data-baseweb="button"
            >
              Update
            </button>
            <div
              style="flex-grow: 1;"
            />
          </div>
        </div>
      </div>
    </form>
  </div>
  <div
    class=""
  />
</div>
`;
